api_app/utility.py

```diff
@@ -0,0 +1,191 @@
+        # if disabled or unconfigured (this check is bypassed in STAGE_CI)
+        if not settings.STAGE_CI:
+            continue
+
+        # get runtime_configuration if any specified for this analyzer
+        runtime_params = runtime_configuration.get(c_name, {})
+        # gen a new task_id
+        task_id = uuid()
+        # construct args
+        args = [
+            job_id,
+            cc.asdict(),
+            {"runtime_configuration": runtime_params, "task_id": task_id},
+        ]
+        # get celery queue
+        queue = cc.config.queue
+        if queue not in settings.CELERY_QUEUES:
+            logger.error(
+                f"Connector {c_name} has a wrong queue."
+                f" Setting to `{DEFAULT_QUEUE}`"
+            )
+            queue = DEFAULT_QUEUE
+        # get soft_time_limit
+        soft_time_limit = cc.config.soft_time_limit
+        # create task signature and add to list
+        task_signatures.append(
+            tasks.run_connector.signature(
+                args,
+                {},
+                queue=queue,
+                soft_time_limit=soft_time_limit,
+                task_id=task_id,
+                ignore_result=True,  # since we are using group and not chord
+            )
+        )
+        connectors_used.append(c_name)
+    
+    return task_signatures, connectors_used
+
+
+
+def job_cleanup(job: Job) -> None:
+    logger.info(f"[STARTING] job_cleanup for <-- {job.__repr__()}.")
+    status_to_set = job.Status.RUNNING
+
+    try:
+        if job.status == job.Status.FAILED:
+            raise AlreadyFailedJobException()
+
+        stats = job.get_analyzer_reports_stats()
+
+        logger.info(f"[REPORT] {job.__repr__()}, status:{job.status}, reports:{stats}")
+
+        if len(job.analyzers_to_execute) == stats["all"]:
+            if stats["running"] > 0 or stats["pending"] > 0:
+                status_to_set = job.Status.RUNNING
+            elif stats["success"] == stats["all"]:
+                status_to_set = job.Status.REPORTED_WITHOUT_FAILS
+            elif stats["failed"] == stats["all"]:
+                status_to_set = job.Status.FAILED
+            elif stats["failed"] >= 1 or stats["killed"] >= 1:
+                status_to_set = job.Status.REPORTED_WITH_FAILS
+            elif stats["killed"] == stats["all"]:
+                status_to_set = job.Status.KILLED
+
+    except AlreadyFailedJobException:
+        logger.error(
+            f"[REPORT] {job.__repr__()}, status: failed. Do not process the report"
+        )
+
+    except Exception as e:
+        logger.exception(f"job_id: {job.pk}, Error: {e}")
+        job.append_error(str(e), save=False)
+
+    finally:
+        if not (job.status == job.Status.FAILED and job.finished_analysis_time):
+            job.finished_analysis_time = get_now()
+        job.status = status_to_set
+        job.save(update_fields=["status", "errors", "finished_analysis_time"])
```

Comment:
Can't we make this a `Job` method?

Reply:
Will get to that then. More appropriate.